loaders/src/glTF/2.0/Extensions/KHR_materials_variants.ts

```diff
@@ -0,0 +1,131 @@
+  material?: Material;
+}
+
+/**
+ * [Specification](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_variants)
+ */
+export class KHR_materials_variants implements IGLTFLoaderExtension {
+  /**
+   * The name of this extension.
+   */
+  public readonly name = NAME;
+
+  private _loader: GLTFLoader;
+
+  public defaultVariant: string | undefined;
+  public availableVariants: string[];
+
+  private _tagsToMap: Map<string, VariantMapping[]> = new Map();
+  
+  constructor(loader: GLTFLoader) {
+    this._loader = loader;
+  }
+
+  public getVariants(): string[] {
+    return Object.keys(this._tagsToMap);
+  }
+
+  public selectVariant(variantName: string | string[]) {
+    if (variantName instanceof Array) {
+      variantName.forEach((name) => this.selectVariantTag(name));
+    } else {
+      this.selectVariantTag(variantName);
+    }
+  }
+
+  public selectVariantTag(variantName: string) {
+    // If the name is valid, switch all meshes to use materials defined by the tags
+    const variantMappings = this._tagsToMap.get(variantName);
+    if (variantMappings === undefined) {
+      return;
+    }
+    variantMappings.forEach((mapping: VariantMapping) => {
+        if (mapping.material) {
+            mapping.mesh.material = mapping.material;
+            return;
+        }
+        mapping.materialPromise.then((material) => {
+            mapping.mesh.material = material;
+        });
+    });
+  }
+
+  /** @hidden */
+  public onLoading(): void {
+    const extensions = this._loader.gltf.extensions;
+    if (extensions && extensions[this.name]) {
+        const extension = extensions[this.name] as IKHRMaterialVariantsTop;
+        this.defaultVariant = extension.default;
+    }
+  }
+
+  public _loadMeshPrimitiveAsync(context: string, name: string, node: INode, mesh: IMesh, primitive: IMeshPrimitive, assign: (babylonMesh: AbstractMesh) => void): Promise<AbstractMesh> {
+    const loadExtensions = GLTFLoader.LoadExtensionAsync<IKHRMaterialVariants, AbstractMesh>(context, primitive, this.name, (extensionContext, extension) => {
+      
+      const assignMesh = (babylonMesh: AbstractMesh) => {
+        assign(babylonMesh);
+        const babylonDrawMode = (GLTFLoader as any)._GetDrawMode(context, primitive.mode);
+        // For each mapping, look at the tags and make a new entry for them
+        extension.mapping.forEach((mapping: IKHRMaterialVariantsMapping) => {
+          mapping.tags.forEach((tag: string, index: number) => {
+            let tag_mapping = this._tagsToMap.get(tag);
+            if (tag_mapping === undefined) {
+                tag_mapping = [];
+            }
+            const material = ArrayItem.Get(`#/materials/`, this._loader.gltf.materials, mapping.material);
+            let meshEntry: VariantMapping = {
+              mesh: babylonMesh,
+              materialPromise: Promise.resolve(null)
+            };
+            if (babylonMesh instanceof Mesh) {
```

Comment:
What happens with instancing? The loader currently uses instancing when multiple nodes point to the same mesh. Does the spec allow material variants on this?

Reply:
Disabled instancing for these meshes. Please check.